contracts/ibc-reflect-send/src/contract.rs

```diff
@@ -0,0 +1,649 @@
+    IbcPacket, IbcReceiveResponse, InitResponse, MessageInfo, Order, QueryResponse, StdError,
+    StdResult,
+};
+
+use crate::msg::{
+    AccountInfo, AccountResponse, AcknowledgementMsg, AdminResponse, BalancesResponse,
+    DispatchResponse, HandleMsg, InitMsg, ListAccountsResponse, PacketMsg, QueryMsg,
+    WhoAmIResponse,
+};
+use crate::state::{accounts, accounts_read, config, config_read, AccountData, Config};
+
+pub const IBC_VERSION: &str = "ibc-reflect-v1";
+
+// TODO: make configurable?
+/// packets live one houe
+const PACKET_LIFETIME: u64 = 60 * 60;
+
+fn build_timeout(block: &BlockInfo) -> Option<u64> {
+    let timeout = block.time + PACKET_LIFETIME;
+    let timeout_nanos = timeout * 1_000_000_000;
+    Some(timeout_nanos)
+}
+
+#[entry_point]
+pub fn init(deps: DepsMut, _env: Env, info: MessageInfo, _msg: InitMsg) -> StdResult<InitResponse> {
+    // we store the reflect_id for creating accounts later
+    let cfg = Config { admin: info.sender };
+    config(deps.storage).save(&cfg)?;
+
+    Ok(InitResponse {
+        messages: vec![],
+        attributes: vec![attr("action", "init")],
+    })
+}
+
+#[entry_point]
+pub fn handle(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::UpdateAdmin { admin } => handle_update_admin(deps, info, admin),
+        HandleMsg::SendMsgs { channel_id, msgs } => {
+            handle_send_msgs(deps, env, info, channel_id, msgs)
+        }
+        HandleMsg::CheckRemoteBalance { channel_id } => {
+            handle_check_remote_balance(deps, env, info, channel_id)
+        }
+        HandleMsg::SendFunds {
+            reflect_channel_id,
+            transfer_channel_id,
+        } => handle_send_funds(deps, env, info, reflect_channel_id, transfer_channel_id),
+    }
+}
+
+pub fn handle_update_admin(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_admin: HumanAddr,
+) -> StdResult<HandleResponse> {
+    // auth check
+    let mut cfg = config(deps.storage).load()?;
+    if info.sender != cfg.admin {
+        return Err(StdError::generic_err("Only admin may set new admin"));
+    }
+    cfg.admin = new_admin;
+    config(deps.storage).save(&cfg)?;
+
+    Ok(HandleResponse {
+        messages: vec![],
+        attributes: vec![
+            attr("action", "handle_update_admin"),
+            attr("new_admin", cfg.admin),
+        ],
+        data: None,
+    })
+}
+
+pub fn handle_send_msgs(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    channel_id: String,
+    msgs: Vec<CosmosMsg>,
+) -> StdResult<HandleResponse> {
+    // auth check
+    let cfg = config(deps.storage).load()?;
+    if info.sender != cfg.admin {
+        return Err(StdError::generic_err("Only admin may send messages"));
+    }
+    // ensure the channel exists (not found if not registered)
+    accounts(deps.storage).load(channel_id.as_bytes())?;
```

Comment:
For what I'm seeing, all the "accounts" (channels) logic is currently being handled "off-contract".

Wouldn't be better (maybe for a future version) if that logic / those requirements (add accounts / connect to channels, validation, address assignment, etc.) are captured in the contract? Either in `init`, or through a dedicated `handle`.

Reply:
Yeah, the tests mock out the external communication and that is all that I would ever do in `cosmwasm` core repo.

In `cosmwasm-plus` I have `multi-test` which would allow the contract-contract calls to be tested (eg. `ibc-reflect` and `reflect`). I could add an `ibc-test` that wraps `multi-test` and allows multi-chain test suites. I have been working with some of those abstractions in Go (and testing this contract there with real IBC stack).